### PR TITLE
Add nightly release docker files

### PR DESCRIPTION
**Pull request by kenny-ezirim**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@ venv
 *.pyc
 test/sagemaker_tests/huggingface/inference/resources/**/*[.json,.h5,.bin,.txt]
 build/
+buildfiles/
```
Comment (saimidu) [inline .gitignore:9]: Why is this necessary?

Comment (kenny-ezirim) [reply .gitignore:9]: This expression will be removed. Needed it to make sure that I don't check in the content of the buildfiles along side with other changes by mistake.

File: .gitignore [hunk 2]
```diff
@@ -5,4 +5,4 @@ venv
 .DS_Store
 *.pyc
 test/sagemaker_tests/huggingface/inference/resources/**/*[.json,.h5,.bin,.txt]
-build/
+build/
```
Comment (saimidu) [inline .gitignore:8]: Could you remove this change so that there is no diff for the `.gitignore` file in this PR?

File: pytorch/training/docker/1.13/py3/example/Dockerfile.sagemaker.lite.gpu [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Expecting base image to be the Deep Learning Container image built by ../cu113/Dockerfile.e3.gpu
+ARG BASE_IMAGE=""
```
Comment (saimidu) [inline pytorch/training/docker/1.13/py3/example/Dockerfile.sagemaker.lite.gpu]: Please go ahead and remove this dockerfile. We do not need it.

File: pytorch/buildspec-sagemaker.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
Comment (saimidu) [inline pytorch/buildspec-sagemaker.yml]: Let's rename this file to buildspec-sagemaker-lite-1-13.yml, and also build sagemaker-lite DLCs in the images specified below.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+# PyTorch Binaries
+ARG PT_E3_TRAINING_URL
+ARG PT_TORCHVISION_URL
+ARG PT_TORCHAUDIO_URL
+ARG PT_TORCHDATA_URL
+ARG PT_S3_WHL
+
+# Install AWS-PyTorch and other torch packages
+RUN pip uninstall -y torch torchvision torchaudio torchdata \
+ && pip install --no-cache-dir -U ${PT_E3_TRAINING_URL} ${PT_TORCHVISION_URL} ${PT_TORCHAUDIO_URL} ${PT_TORCHDATA_URL}
```
Comment (saimidu) [inline pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu:179]: We want to use PyTorch RCs or nightly from the pytorch conda channel, instead of the AWS binaries.

File: src/image_builder.py [hunk 6]
```diff
@@ -138,7 +168,7 @@ def image_builder(buildspec):
                 }
             }
         )
-
+
```
Comment (saimidu) [inline src/image_builder.py:171]: nit: Please remove trailing whitespace. This is possibly an IDE configuration that can be used to automatically delete trailing whitespace in a file when it is saved.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu [hunk 7]
```diff
@@ -148,50 +142,47 @@ RUN pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pytho
...
- && pip install --no-cache-dir -U ${PT_E3_TRAINING_URL} ${PT_TORCHVISION_URL} ${PT_TORCHAUDIO_URL} ${PT_TORCHDATA_URL}
+# uninstall if already present
+RUN pip uninstall -y torch torchvision torchaudio torchdata
 
-# Install PT S3 plugin
-RUN pip install -U ${PT_S3_WHL}
-RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
+# TODO: Install AWS-PyTorch and other torch packages here
+
+RUN /opt/conda/bin/conda install -c conda-forge \
```
Comment (saimidu) [inline pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu:175]: This should be `-c pytorch-nightly` -- https://anaconda.org/pytorch-nightly/repo

File: pytorch/training/docker/1.13/py3/example/Dockerfile.sagemaker.lite.gpu (hunk 3 above)
Comment (kenny-ezirim) [reply pytorch/training/docker/1.13/py3/example/Dockerfile.sagemaker.lite.gpu]: The examples are gone.

File: pytorch/buildspec-sagemaker.yml (hunk 4 above)
Comment (kenny-ezirim) [reply pytorch/buildspec-sagemaker.yml]: Renamed.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu (hunk 7 above)
Comment (kenny-ezirim) [reply pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu]: Changed.

File: pytorch/buildspec-sagemaker-lite-1-13.yml [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  BuildSageMakerGPUPTTrainPy3DockerImage:
+    <<: *TRAINING_REPOSITORY
+    build: &PYTORCH_GPU_TRAINING_PY3 false
+    image_size_baseline: 14000
+    device_type: &DEVICE_TYPE gpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py38
+    cuda_version: &CUDA_VERSION cu113
+    os_version: &OS_VERSION ubuntu20.04
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-sagemaker" ]
```
Comment (saimidu) [inline pytorch/buildspec-sagemaker-lite-1-13.yml:71]: Please change this to end in `"-sagemaker-lite"`

File: pytorch/buildspec-sagemaker-lite-1-13.yml [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  # Build Sagemaker CPU Training Image
+  BuildSageMakerCPUPTTrainPy3DockerImage:
+    <<: *TRAINING_REPOSITORY
+    build: &PYTORCH_CPU_TRAINING_PY3 false
+    image_size_baseline: 5000
+    device_type: &DEVICE_TYPE cpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py38
+    os_version: &OS_VERSION ubuntu20.04
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-sagemaker" ]
```
Comment (saimidu) [inline pytorch/buildspec-sagemaker-lite-1-13.yml:53]: Please change this to end in `"-sagemaker-lite"`

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+ENV SAGEMAKER_TRAINING_MODULE=sagemaker_pytorch_container.training:main
+
+# TODO: Redeclare ARGs for PyTorch and other Torch packages here
+
+# uninstall if already present
+RUN pip uninstall -y torch torchvision torchaudio torchdata
+
+# TODO: Install AWS-PyTorch and other torch packages here
+
+RUN /opt/conda/bin/conda install -c -c pytorch-nightly \
```
Comment (saimidu) [inline pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu:222]: There are two `-c`s here.

File: src/image_builder.py [hunk 11]
```diff
@@ -80,10 +102,18 @@ def image_builder(buildspec):
 
         if image_config.get("context") is not None:
             ARTIFACTS.update(image_config["context"])
-
+
```
Comment (saimidu) [inline src/image_builder.py:105]: Please remove trailing whitespace.

File: src/image_builder.py [hunk 12]
```diff
@@ -268,25 +297,26 @@ def image_builder(buildspec):
     # From all images, filter the images that were supposed to be built and upload their metrics
     BUILT_IMAGES = [image for image in ALL_IMAGES if image.to_build]
 
-    FORMATTER.banner("Upload Metrics")
-    upload_metrics(BUILT_IMAGES, BUILDSPEC, is_any_build_failed, is_any_build_failed_size_limit)
+    if BUILT_IMAGES:
```
Comment (saimidu) [inline src/image_builder.py:300]: Good catch.

**Review by saimidu — COMMENTED** [review]
Please remove the TF 2.9 sm lite and full dockerfiles.

File: pytorch/buildspec-sagemaker-lite-1-13.yml (hunk 9 above)
Comment (kenny-ezirim) [reply pytorch/buildspec-sagemaker-lite-1-13.yml:53]: Fixed.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu (hunk 10 above)
Comment (kenny-ezirim) [reply pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.lite.cpu:222]: Fixed.

File: pytorch/training/docker/1.13/py3/cu113/Dockerfile.sagemaker.lite.gpu [hunk 13]
```diff
@@ -228,10 +228,13 @@ ARG PYTHON_SHORT_VERSION
...
-RUN /opt/conda/bin/conda install -c pytorch-nightly torchvision 
-RUN /opt/conda/bin/conda install -c pytorch-nightly torchaudio 
-RUN /opt/conda/bin/conda install -c pytorch-nightly torchdata
+RUN /opt/conda/bin/conda install \
+   pytorch=1.13.* \
+   torchvision \
+   torchaudio \
+   torchdata \
+   cudatoolkit=11.3 \
+   -c pytorch-nightly
```
Comment (saimidu) [inline pytorch/training/docker/1.13/py3/cu113/Dockerfile.sagemaker.lite.gpu:237]: Why move this to the end, rather than specifying it at the beginning of the command like the rest of the dockerfile? Does it help with something?

File: test/testrunner.py [hunk 14]
```diff
@@ -381,6 +382,12 @@ def main():
             cmd_exit_statuses = [pytest.main(pytest_cmd) for pytest_cmd in pytest_cmds]
             if all([status == 0 for status in cmd_exit_statuses]):
                 sys.exit(0)
+            elif any([status != 0 for status in cmd_exit_statuses]) and is_nightly_context():
+                LOGGER.warning("\nSuppresed Failed Nightly Tests")
+                for index, status in enumerate(cmd_exit_statuses):
+                    if status != 0:
+                        LOGGER.warning(f'"{test_path}" tests failed. Status code: {status}')
```
Comment (saimidu) [inline test/testrunner.py:389]: Should `test_path` be `pytest_cmds[index]`?

`test_path` is not valid here.

**Review by saimidu — COMMENTED** [review]
Please remove the TF 2.9 sm lite/full dockerfiles from this PR.

**Comment by saimidu** [comment]
Please also enable the SM tests.

File: test/sagemaker_tests/tensorflow/tensorflow1_training/integration/conftest.py [hunk 15]
```diff
@@ -24,6 +24,14 @@
 from sagemaker.tensorflow import TensorFlow
 from ..integration import NO_P2_REGIONS, NO_P3_REGIONS, NO_P4_REGIONS, get_ecr_registry
 
+from test.test_utils import (
+    is_nightly_context,
+    NightlyFeatureLabel
+)
+from test.test_utils.imageutils import (
+    are_fixture_labels_enabled
+)
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/sagemaker_tests/tensorflow/tensorflow1_training/integration/conftest.py:33]: Please add these in the same way as done in the PyTorch Training conftest.py.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.full.cpu [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+COPY changehostname.c /
+COPY start_with_right_hostname.sh /usr/local/bin/start_with_right_hostname.sh
+
+RUN chmod +x /usr/local/bin/start_with_right_hostname.sh
+
+# Install scikit-learn and pandas
+RUN conda install -y -c conda-forge \
+    scikit-learn \
+    pandas \
+    openssl==1.1.1o
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.full.cpu:259]: It can't be more general because there are multiple versions of openssl with different major versions. The current version in the image that is prone to vulnerabilities is 1.1.1q. Version 1.1.1o is the only version that addresses all 3 current CVEs that was raised. Attempts to do it at the same level as pyopenssl failed as the version 1.1.1q could not be overwritten with the latest version.

File: test/dlc_tests/sanity/test_pre_release.py [hunk 17]
```diff
@@ -355,11 +363,18 @@ def test_framework_and_cuda_version_gpu(gpu, ec2_connection):
                 version_to_check = "0.3.1" if tag_framework_version == "0.3.2" else tag_framework_version
                 assert output.stdout.strip().startswith(version_to_check)
             elif tested_framework == "torch" and Version(tag_framework_version) >= Version("1.10.0"):
-                torch_version_pattern = r"{torch_version}(\+cu\d+)".format(torch_version=tag_framework_version)
-                assert re.fullmatch(torch_version_pattern, output.stdout.strip()), (
-                    f"torch.__version__ = {output.stdout.strip()} does not match {torch_version_pattern}\n"
-                    f"Please specify framework version as X.Y.Z+cuXXX"
-                )
+                if is_nightly_context():
+                    torch_version_pattern = r"{torch_version}(\.dev\d+)".format(torch_version=tag_framework_version)
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/dlc_tests/sanity/test_pre_release.py:367]: Yes, does not have +cuXXX.

File: pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.full.cpu (hunk 16 above)
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply pytorch/training/docker/1.13/py3/Dockerfile.sagemaker.full.cpu:259]: Removed.

File: test/test_utils/sagemaker.py [hunk 18]
```diff
@@ -314,6 +314,7 @@ def execute_local_tests(image, pytest_cache_params):
         ec2_conn.run(f"tar -xzf {sm_tests_tar_name}")
         kill_background_processes_and_run_apt_get_update(ec2_conn)
         with ec2_conn.cd(path):
+            print(f"Changed path to: {path}")
```
Comment (saimidu) [inline test/test_utils/sagemaker.py:317]: Is this a temporary addition, or is it meant to be permanent? If permanent, please use `LOGGER` instead of `print` statements.

Comment (kenny-ezirim) [reply test/test_utils/sagemaker.py:317]: Temporal. I want to see the directory switched to.

File: test/test_utils/sagemaker.py [hunk 19]
```diff
@@ -332,17 +333,20 @@ def execute_local_tests(image, pytest_cache_params):
...
-                ec2_conn.get(ec2_test_report_path, os.path.join("test", f"{job_type}_{tag}_sm_local.xml"))
-    except:
-        print(f"Exception {sys.exc_info()[0]} occurred")
+                try:
+                    ec2_conn.run(pytest_command)
+                    print(f"Downloading Test reports for image: {image}")
+                    ec2_conn.get(ec2_test_report_path, os.path.join("test", f"{job_type}_{tag}_sm_local.xml"))
+                except:
+                    if not is_nightly_context():
+                        raise ValueError(f"Sagemaker Local tests failed for {image}")
```
Comment (saimidu) [inline test/test_utils/sagemaker.py:345]: Please raise the same exception as the one from the `try` block, rather than raising a new `ValueError`, because this breaks the flow of the Traceback.

For example:
```
try:
    ...
except Exception as e:
    if is_nightly_context():
        LOGGER.warning(f"SageMaker Local tests failed for {image} with exception: {e}")
    else:
        raise e
```

Comment (saimidu) [reply test/test_utils/sagemaker.py:345]: Also, please add the `except:` block back into the outer `try-except-finally` block.

```
            else:
                try:
                    ec2_conn.run(pytest_command)
                    ...
                except Exception as e:
                    ....
    except:
        print(f"Exception {sys.exc_info()[0]} occurred")
```

Comment (kenny-ezirim) [reply test/test_utils/sagemaker.py:345]: Do you want to bring in the logging functionality into this module? Up until now there was no logging mechanism.

File: test/test_utils/sagemaker.py [hunk 20]
```diff
@@ -332,13 +334,20 @@ def execute_local_tests(image, pytest_cache_params):
...
+                            raise ValueError(f"Sagemaker Local tests failed for {image}")
             else:
-                ec2_conn.run(pytest_command)
-                print(f"Downloading Test reports for image: {image}")
-                ec2_conn.get(ec2_test_report_path, os.path.join("test", f"{job_type}_{tag}_sm_local.xml"))
-    except:
-        print(f"Exception {sys.exc_info()[0]} occurred")
+                try:
+                    ec2_conn.run(pytest_command)
+                    print(f"Downloading Test reports for image: {image}")
```
Comment (saimidu) [inline test/test_utils/sagemaker.py:344]: As discussed on call, please go ahead and change this to a `warn=True` and use the output Result object to figure out further steps.

File: test/test_utils/sagemaker.py [hunk 21]
```diff
@@ -368,6 +377,7 @@ def execute_sagemaker_remote_tests(process_index, image, global_pytest_cache, py
     account_id = os.getenv("ACCOUNT_ID", boto3.client("sts").get_caller_identity()["Account"])
     pytest_cache_util = PytestCache(boto3.client("s3"), account_id)
     pytest_command, path, tag, job_type = generate_sagemaker_pytest_cmd(image, SAGEMAKER_REMOTE_TEST_TYPE)
+    test_python_path = os.path.join(UBUNTU_HOME_DIR, "test")
```
Comment (saimidu) [inline test/test_utils/sagemaker.py:380]: This is no longer used, and can therefore be removed, right?

File: test/testrunner.py [hunk 22]
```diff
@@ -83,6 +84,7 @@ def run_sagemaker_test_in_executor(image, num_of_instances, instance_type):
         log_return.update_pool("running", instance_type, num_of_instances, job_type)
         context = Context()
         with context.cd(path):
+            print(f"Changed path to: {path}")
```
Comment (saimidu) [inline test/testrunner.py:87]: Please remove this, or change to `LOGGER.info`.

**Review by saimidu — COMMENTED** [review]
Please remove the TF 2.9 sm lite/full dockerfiles.

**Review by saimidu — DISMISSED** [review]
Approved if SM tests pass.

Let's also wait for the results of all tests on PT and TF 2 after disabling `nightly_pr_test_mode` to ensure that the new changes haven not affected testing for non-nightly images.

**Review by saimidu — APPROVED** [review]

**Review by Lokiiiiii — DISMISSED** [review]

**Review by mariumof — APPROVED** [review]

**Review by saimidu — DISMISSED** [review]

**Review by saimidu — APPROVED** [review]
